Fix for incorrect replacing with phrases with several spaces inside

Comandeer opened the pull request:
<!--
🚨 If you want to submit a PR for a security vulnerability, please contact us directly
at https://ckeditor.com/contact/ instead. 🚨
-->
## What is the purpose of this pull request?

Bug fix

## Does your PR contain necessary tests?

All patches that change the editor code must include tests. You can always read more
on [PR testing](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_contributing_code.html#tests),
[how to set the testing environment](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_tests.html) and
[how to create tests](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_tests.html#creating-your-own-test)
in the official CKEditor documentation.

### This PR contains

- [x] Unit tests
- [x] Manual tests

## Did you follow the CKEditor 4 code style guide?

Your code should follow the guidelines from the [CKEditor 4 code style guide](https://github.com/ckeditor/ckeditor4/blob/major/dev/docs/codestyle.md) which helps keep the entire codebase consistent.

- [x] PR is consistent with the code style guide

## What is the proposed changelog entry for this pull request?

```
* [#5061](https://github.com/ckeditor/ckeditor4/issues/5061): Fixed: [Find / Replace](https://ckeditor.com/cke4/addon/find) plugin incorrectly handles multiple whitespace during replacing text.
```

## What changes did you make?

Creating a text node from any text removes any consecutive whitespace. To prevent that I've added a `replace()` that replaced every odd space with a non-breaking one.

The only concern I have is if we want to replace all spaces with the non-breaking one or only "normal" spaces (` `)? For now the code replaces all of them.

## Which issues does your PR resolve?

Closes #5061.
<!-- Closes #<ANOTHER_ISSUE_NUMBER>. -->

Comandeer commented:
It seems that this PR should be merged after #5066 as the other one is fixing failing `find` tests.

Comandeer commented:
Rebased onto latest `master`.

KarolDawidziuk commented on plugins/find/dialogs/find.js:
```diff
@@ -486,8 +488,17 @@
-					var domRange = this.matchRange.toDomRange();
-					var text = editor.document.createText( newString );
+					var domRange = this.matchRange.toDomRange(),
+						textWithPreservedSpaces = newString.replace( consecutiveWhitespaceRegex,
+							function( whitespace ) {
+								var newSpaces = CKEDITOR.tools.array.map( whitespace, function( space, i ) {
+									return i % 2 === 0 ? nonBreakingSpace : space;
+								} );
+
+								return newSpaces.join( '' );
+							} ),
```
Overall code works good but let's move the replace logic into a separate function. It will look clearer.

KarolDawidziuk commented on tests/plugins/find/find.js:
```diff
@@ -263,6 +263,41 @@ bender.test( {
+		bot.dialog( 'replace', function( dialog ) {
+			dialog.setValueOf( 'replace', 'txtReplace', 'foo   bar' );
+			dialog.getContentElement( 'replace', 'btnReplaceAll' ).click();
+			dialog.getButton( 'cancel' ).click();
```
We can remove this line because we hide dialog after assertion. ( line 297 )

Comandeer replied:
Actually, I've moved it before the assertion in both tests. Otherwise, if the assertion failed, the dialog wouldn't be closed.

KarolDawidziuk reviewed (changes requested):
Hi, @Comandeer 
Overall everything is good and your fix works well.

There are some small style adjustments to add. Please see my inline comments.

Comandeer commented on tests/plugins/find/find.js:
```diff
@@ -263,6 +263,41 @@ bender.test( {
+			dialog.getContentElement( 'replace', 'btnReplaceAll' ).click();
+			dialog.getButton( 'cancel' ).click();
+
+			assert.areSame( '<p>foo&nbsp; &nbsp;bar</p><p>foo&nbsp; &nbsp;bar</p>', bot.getData( false, true ) );
```
I've just copied the code from some previous tests :see\_no\_evil: I'll remove unnecessary parameters.

Comandeer commented on tests/plugins/find/manual/replacedoublespace.html:
```diff
@@ -0,0 +1,9 @@
+
+<script>
+	CKEDITOR.replace( 'editor', {
+		language: 'en'
```
No, it's not. Leftover from testing in my Chrome that makes the editor using Polish :upside\_down\_face:

Comandeer commented:
I've moved the logic into a separate function and updated tests – switching also from "Replace" to "Replace All" (as "Replace" need two clicks and "Replace All" only one) in them. Also rebased the PR onto latest `master`.

KarolDawidziuk reviewed (approved):
Thanks, @Comandeer for the improvements.
For me, everything looks good, so it's R+ from my side. 🎉

jacekbogdanski commented on plugins/find/dialogs/find.js:
```diff
@@ -826,6 +829,19 @@
+		function createTextNodeWithPreservedSpaces( editor, text ) {
+			var textWithPreservedSpaces = text.replace( consecutiveWhitespaceRegex,
+				function( whitespace ) {
+					var newSpaces = CKEDITOR.tools.array.map( whitespace, function( space, i ) {
```
`CKEDITOR.tools.array.map` function takes an array as an argument, not a string. This code takes advantage of strings defining iterator, but it reduces code readability.

jacekbogdanski reviewed (changes requested):
We are missing a case when text includes additional space at the end of a phrase, e.g. `foobar<space>`. I've added a test in https://github.com/ckeditor/ckeditor4/pull/5065/commits/76d3d0131a3dc31fbb3b72221367a91bb6c32e07 confirming that issue.

Comandeer commented:
> We are missing a case when text includes additional space at the end of a phrase

Spaces at the end of phrases are preserved as long as the end of the phrase is not the same as the end of the paragraph. In such a case it seems to be a totally separate issue, connected with how `CKEDITOR.dom.range#insertNode()` works (see #5074). In that case, I propose to simply remove the failing test and resolve that issue separately.

jacekbogdanski commented:
Alright, thanks for checking this out and extracting a follow-up ticket. Let's wait for CI and if it's green we can merge this PR 🎉
